### PR TITLE
[ISSUE #3367] optimeze selector and rule match 

### PR DESCRIPTION
**Pull request by afumu**
optimeze selector and rule match

### CODE REVIEW

File: shenyu-bootstrap/src/main/resources/application.yml [hunk 1]
```diff
@@ -64,6 +64,9 @@ management:
       enabled: false
 
 shenyu:
+  matchCache:
+    enabled: true
+    maxFreeMemory:
```
Comment (dragon-zhang) [inline shenyu-bootstrap/src/main/resources/application.yml:69]: should add a default value.

File: shenyu-plugin/shenyu-plugin-base/src/main/java/org/apache/shenyu/plugin/base/cache/MatchDataCache.java [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+     * @param path       the path
+     * @return the selector data
+     */
+    public List<SelectorData> obtainSelectorData(final String pluginName, final String path) {
+        List<SelectorData> selectorDataList = null;
+        final LRUMap<String, List<SelectorData>> lruMap = SELECTOR_DATA_MAP.get(pluginName);
+        if (Objects.nonNull(lruMap)) {
+            selectorDataList = lruMap.get(path);
+        }
+        return selectorDataList;
```
Comment (dragon-zhang) [inline shenyu-plugin/shenyu-plugin-base/src/main/java/org/apache/shenyu/plugin/base/cache/MatchDataCache.java:240]: please use `java.util.Optional`

File: shenyu-plugin/shenyu-plugin-base/src/main/java/org/apache/shenyu/plugin/base/AbstractShenyuPlugin.java [hunk 3]
```diff
@@ -63,38 +73,133 @@ public abstract class AbstractShenyuPlugin implements ShenyuPlugin {
...
+            }
+
+        }
+        return Pair.of(Boolean.TRUE, null);
+    }
+
+    private void cacheSelectorDataIfEnabled(final String path, final SelectorData selectorData) {
+        if (matchCacheConfig.getEnabled()) {
+            if (Objects.isNull(selectorData)) {
+                nullSelectorData.setPluginName(named());
```
Comment (dragon-zhang) [inline shenyu-plugin/shenyu-plugin-base/src/main/java/org/apache/shenyu/plugin/base/AbstractShenyuPlugin.java:169]: Here has a concurrency problem, we have many plugins.

File: shenyu-plugin/shenyu-plugin-base/src/main/java/org/apache/shenyu/plugin/base/AbstractShenyuPlugin.java [hunk 4]
```diff
@@ -42,6 +46,12 @@ public abstract class AbstractShenyuPlugin implements ShenyuPlugin {
 
     private static final Logger LOG = LoggerFactory.getLogger(AbstractShenyuPlugin.class);
 
+    private static ShenyuConfig.MatchCache matchCacheConfig;
+
+    private static SelectorData nullSelectorData = new SelectorData();
+
+    private static RuleData nullRuleData = new RuleData();
```
Comment (dragon-zhang) [inline shenyu-plugin/shenyu-plugin-base/src/main/java/org/apache/shenyu/plugin/base/AbstractShenyuPlugin.java:53]: Oh, `nullSelectorData` and `nullRuleData`, maybe shouldn not use `static` modifer?

File: shenyu-plugin/shenyu-plugin-base/src/main/java/org/apache/shenyu/plugin/base/cache/MatchDataCache.java [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+     * Cache rule data.
+     *
+     * @param path      the path
+     * @param ruleData  the rule data
+     * @param maxMemory the max memory
+     */
+    public void cacheRuleData(final String path, final RuleData ruleData, final Integer maxMemory) {
+        final LRUMap<String, List<RuleData>> lruMap = RULE_DATA_MAP.computeIfAbsent(ruleData.getPluginName(),
+            map -> new MemorySafeLRUMap<>(maxMemory, 1 << 16));
+        List<RuleData> ruleDataList = lruMap.computeIfAbsent(path, list -> Collections.synchronizedList(new ArrayList<>()));
```
Comment (dragon-zhang) [inline shenyu-plugin/shenyu-plugin-base/src/main/java/org/apache/shenyu/plugin/base/cache/MatchDataCache.java:220]: Here may be also use `Collections.synchronizedList(new LinkedList<>())`.

**Comment by zouchangfu** [comment]
I will close this pr ,pls see this [https://github.com/apache/incubator-shenyu/pull/3440](https://github.com/apache/incubator-shenyu/pull/3440)

**Pull request closed without merging**